src/ameba.ts

```diff
@@ -27,23 +28,68 @@ export class Ameba {
         this.config = getConfig();
     }
 
-    public execute(document: TextDocument): void {
+    public execute(document: TextDocument, virtual = false): void {
         if (document.languageId !== 'crystal' || document.isUntitled || document.uri.scheme !== 'file') {
             return;
         }
 
-        const args = [this.config.command, document.fileName, '--format', 'json'];
-        const dir = workspace.getWorkspaceFolder(document.uri)!.uri.fsPath;
-        const configFile = path.join(dir, this.config.configFileName);
+        // When reading from stdin, cannot pass any files via the CLI, otherwise there will be overlaps
+        // and potentially out of date errors
+        const devNull = process.platform === "win32" ? "nul" : "/dev/null"
+        const mainFile = virtual ? devNull : document.fileName
+
+        const args = [this.config.command, mainFile, '--format', 'json'];
+        const space = workspace.getWorkspaceFolder(document.uri) ?? noWorkspaceFolder(document)
+
+        const configFile: string = path.join(space.uri.fsPath, this.config.configFileName);
+
         if (existsSync(configFile)) args.push('--config', configFile);
 
+        if (virtual) {
+            args.push('--stdin-filename', document.uri.fsPath)
+
+            // Disabiling these as they're common when typing
+            args.push("--except=Lint/Formatting,Layout/TrailingBlankLines,Layout/TrailingWhitespace")
+        }
+
```

Comment:
No need to pass any filename when reading from STDIN.

```suggestion
        const space = workspace.getWorkspaceFolder(document.uri) ?? noWorkspaceFolder(document);
        const args = [this.config.command];

        if (!virtual) {
            args.push(document.fileName);
        } else {
            args.push('--stdin-filename', document.uri.fsPath);

            // Disabling these as they're common when typing
            args.push('--except', 'Lint/Formatting,Layout/TrailingBlankLines,Layout/TrailingWhitespace');
        }
        args.push('--format', 'json');

        const configFile = path.join(space.uri.fsPath, this.config.configFileName);
        if (existsSync(configFile)) args.push('--config', configFile);

```